src/category_theory/sites/dense_subsite.lean

```diff
@@ -0,0 +1,388 @@
+lemma pushforward_family_compatible {X} (x : ℱ.val.obj (op X)) :
+  (pushforward_family H α x).compatible :=
+begin
+  intros Y₁ Y₂ Z g₁ g₂ f₁ f₂ h₁ h₂ e,
+  apply H.sheaf_ext,
+  intros Y f,
+  simp only [pushforward_family, ←functor_to_types.map_comp_apply, ←op_comp],
+  change (ℱ.val.map _ ≫ α.app (op _) ≫ ℱ'.val.map _) _ =
+    (ℱ.val.map _ ≫ α.app (op _) ≫ ℱ'.val.map _) _,
+  rw ← G.image_preimage (f ≫ g₁ ≫ _),
+  rw ← G.image_preimage (f ≫ g₂ ≫ _),
+  erw ← α.naturality (G.preimage _).op,
+  erw ← α.naturality (G.preimage _).op,
+  refine congr_fun _ x,
+  simp only [quiver.hom.unop_op, functor.comp_map, ←op_comp, ←category.assoc,
+    functor.op_map, ←ℱ.val.map_comp, G.image_preimage],
+  congr' 3,
+  simp[e]
+end
+
+/-- (Implementation). The morphism `ℱ(X) ⟶ ℱ'(X)` given by glueing the `pushforward_family`. -/
+noncomputable
+def app_hom (X : D) : ℱ.val.obj (op X) ⟶ ℱ'.val.obj (op X) := λ x,
+  (ℱ'.property _ (H.is_cover X)).amalgamate
+    (pushforward_family H α x)
+    (pushforward_family_compatible H α x)
+
+@[simp] lemma pushforward_family_apply {X} (x : ℱ.val.obj (op X)) {Y : C} (f : G.obj Y ⟶ X) :
+  pushforward_family H α x f (presieve.in_cover_by_image G f) = α.app (op Y) (ℱ.val.map f.op x) :=
+begin
+  unfold pushforward_family,
+  refine congr_fun _ x,
+  rw ←G.image_preimage (get_cover_by_image_structure _).lift,
+  change ℱ.val.map _ ≫ α.app (op _) ≫ ℱ'.val.map _ = ℱ.val.map f.op ≫ α.app (op Y),
+  erw ←α.naturality (G.preimage (get_cover_by_image_structure _).lift).op,
+  simp only [←functor.map_comp, ←category.assoc, functor.comp_map, G.image_preimage,
+     G.op_map, quiver.hom.unop_op, ←op_comp, presieve.cover_by_image_structure.fac],
+end
+
+@[simp] lemma app_hom_restrict {X : D} {Y : C} (f : op X ⟶ op (G.obj Y)) (x) :
+  ℱ'.val.map f (app_hom H α X x) = α.app (op Y) (ℱ.val.map f x) :=
+begin
+  refine ((ℱ'.property _ (H.is_cover X)).valid_glue
+    (pushforward_family_compatible H α x) f.unop (presieve.in_cover_by_image G f.unop)).trans _,
+  apply pushforward_family_apply
+end
+
+@[simp] lemma app_hom_valid_glue {X : D} {Y : C} (f : op X ⟶ op (G.obj Y)) :
+  app_hom H α X ≫ ℱ'.val.map f = ℱ.val.map f ≫ α.app (op Y) :=
+by { ext, apply app_hom_restrict }
+
+/--
+(Implementation). The maps given in `app_iso` is inverse to each other and gives a `ℱ(X) ≅ ℱ'(X)`.
+-/
+@[simps] noncomputable
+def app_iso (i : G.op ⋙ ℱ.val ≅ G.op ⋙ ℱ'.val) (X : D) :
+  ℱ.val.obj (op X) ≅ ℱ'.val.obj (op X) :=
+{ hom := app_hom H i.hom X,
+  inv := app_hom H i.inv X,
+  hom_inv_id' :=
+  begin
+    ext x,
+    apply H.sheaf_ext,
+    intros Y f,
+    simp
+  end,
+  inv_hom_id' :=
+  begin
+    ext x,
+    apply H.sheaf_ext,
+    intros Y f,
+    simp
+  end }
+
+/--
+Given an natural transformation `G ⋙ ℱ ⟶ G ⋙ ℱ'` between sheaves of types, we may obtain a
+natural transformation between sheaves.
+-/
+@[simps] noncomputable
+def sheaf_hom (α : G.op ⋙ ℱ.val ⟶ G.op ⋙ ℱ'.val) : ℱ.val ⟶ ℱ'.val :=
```

Comment:
This builds a morphism of presheaves, rather than a morphism of sheaves right? Is there a good reason not to upgrade (which I think comes for free?)

Reply:
I suppose that morphisms of presheaves and sheaves are the same thing to Lean, so this could also be used to obtain a morphism of sheaves. The reason I defined it as a morphism of presheaves is because `α` is also a morphism of presheaves.

I could change the type ascription if you think it is better that way.

Reply:
I think the type checker will find it easier to treat a morphism of sheaves as a morphism of presheaves than vice versa.

Reply:
Anyway, one of the sheaves have been changed to a presheaf to make the statement more general :)